src/main/cpp/util/profiler.h

```diff
@@ -0,0 +1,158 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BAZEL_SRC_MAIN_CPP_UTIL_PROFILER_H_
+#define BAZEL_SRC_MAIN_CPP_UTIL_PROFILER_H_
+
+#include <stdint.h>  // int64_t
+#include <stdio.h>   // printf
+
+namespace blaze_util {
+namespace profiler {
+
+// A time duration, measured in some implementation-dependent units.
+//
+// Using a struct to wrap int64_t yields a unique type that we can't
+// accidentally use as a type-synonym of other int64_t types.
+typedef struct _Ticks {
+  int64_t value_;
+  static _Ticks Now();
+} Ticks;
+
+// A time duration, measured in milliseconds.
+//
+// Using a struct to wrap int64_t yields a unique type that we can't
+// accidentally use as a type-synonym of other int64_t types.
+typedef struct _Duration {
+  int64_t millis_;
+  static _Duration FromTicks(const Ticks ticks);
+} Duration;
+
+// Collects statistics about a function or some C++ scope.
+//
+// Usage: see ScopeProfiler.
+//
+// Records the number of times the scope was entered (the function called) and
+// the total time spent in there.
+class Stats {
+ protected:
+  int32_t calls_;
+  Ticks total_;
+
+ public:
+  Stats() : calls_(0), total_({0}) {}
+  virtual ~Stats() {}
+  void AddCall() { calls_++; }
+  void AddTicks(const Ticks t) { total_.value_ += t.value_; }
+  uint32_t GetCalls() const { return calls_; }
+  Duration GetDuration() const { return Duration::FromTicks(total_); }
+};
+
+// Collects `Stats` and prints them in its destructor.
+//
+// Usage: see ScopeProfiler.
+class PrintingStats {
+  const char* name_;
+  Stats stats_;
+
+ public:
+  // Constructor.
+  // Args:
+  //   name: a descriptive name, will be printed in the output
+  PrintingStats(const char* name) : name_(name) {}
+  ~PrintingStats();
+
+  Stats* GetStats() { return &stats_; }
+};
+
+// Measures elapsed time.
+class StopWatch {
+  Ticks start_;
+
+ public:
+  // Constructor -- it also resets this StopWatch.
+  StopWatch() { start_ = Ticks::Now(); }
+
+  virtual ~StopWatch() {}
+
+  // Returns the elapsed time since the last reset.
```

Comment:
nit: add "the elapsed time in implementation-defined ticks". Also, add docstring for `ElapsedTime()`

Reply:
Done.